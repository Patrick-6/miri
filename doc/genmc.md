# **(WIP)** Documentation for Miri-GenMC
[GenMC](https://github.com/MPI-SWS/genmc) is a stateless model checker for exploring concurrent executions of a program.

**NOTE: Currently, no actual GenMC functionality is part of Miri, this is still WIP.**

<!-- FIXME(genmc): add explanation. -->

## Usage
Basic usage:
```shell
MIRIFLAGS="-Zmiri-genmc" cargo miri run
```

<!-- FIXME(genmc): explain options. -->

<!-- FIXME(genmc): explain Miri-GenMC specific functions. -->

## Tips

<!-- FIXME(genmc): add tips for using Miri-GenMC more efficiently. -->

## Limitations

Some or all of these limitations might get removed in the future:

- Borrow tracking is currently incompatible (stacked/tree borrows).
- Only Linux is supported for now.
- No 32-bit platform support.
- No cross-platform interpretation.

<!-- FIXME(genmc): document remaining limitations -->

## Development

GenMC is written in C++, which complicates development a bit.
For Rust-C++ interop, Miri uses [CXX.rs](https://cxx.rs/), and all handling of C++ code is contained in the `genmc-sys` crate (located in the Miri repository root directory: `miri/genmc-sys/`).

Building GenMC requires a compiler with C++23 support.
<!-- FIXME(genmc,llvm): remove once LLVM dependency is no longer required. -->
Currently, building GenMC also requires linking to LLVM, which needs to be installed manually.

The actual code for GenMC is not contained in the Miri repo itself, but in a [separate GenMC repo](https://github.com/MPI-SWS/genmc) (with different maintainers).
Note that this repo is just a mirror repo.
<!-- FIXME(genmc): define how submitting code to GenMC should be handled. -->

<!-- FIXME(genmc): explain development. -->

### Building the GenMC Library
The build script in the `genmc-sys` crate handles locating, downloading, building and linking the GenMC library.

To determine which GenMC repo path will be used, the following steps are taken:
- If the env var `GENMC_SRC_PATH` contains a path to a directory with a GenMC repo, use that path.
- If the path `genmc-sys/genmc` exists, use that path.
- If the path `genmc-sys/downloaded/genmc` exists, try to set the GenMC repo there to the commit we need.
- If the downloaded repo doesn't exist or is missing the commit, the build script will fetch the commit over the network.
  - Note that the build script will *not* access the network if any of the steps previous steps succeeds.

Once we get the path to the repo, the compilation proceeds in two steps:
- Compile GenMC into a library (using cmake).
- Compile the cxx.rs bridge to connect the library to the Rust code.
The first step is where all build settings are made, the relevant ones are then stored in a `config.h` file that can be included in the second compilation step.
