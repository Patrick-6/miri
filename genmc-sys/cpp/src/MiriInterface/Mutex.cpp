/** This file contains functionality related to handling mutexes.  */

#include "MiriInterface.hpp"

// CXX.rs generated headers:
#include "genmc-sys/src/lib.rs.h"

auto MiriGenmcShim::handle_mutex_lock(ThreadId thread_id, uint64_t address, uint64_t size)
    -> MutexLockResult {
    // TODO GENMC: this needs to be identical even in multithreading
    ModuleID::ID annot_id;
    if (annotation_id.contains(address)) {
        annot_id = annotation_id.at(address);
    } else {
        annot_id = annotation_id_counter++;
        annotation_id.insert(std::make_pair(address, annot_id));
    }
    const auto aSize = ASize(size);
    auto annot = std::move(Annotation(
        AssumeType::Spinloop,
        Annotation::ExprVP(NeExpr<AnnotID>::create(
                               RegisterExpr<AnnotID>::create(aSize.getBits(), annot_id),
                               ConcreteExpr<AnnotID>::create(aSize.getBits(), SVal(1))
        )
                               .release())
    ));

    // Mutex starts out unlocked, so we always say the previous value is "unlocked".
    auto old_val_setter = [this](SAddr addr) {
        this->handle_old_val(addr, GenmcScalarExt::from_sval(SVal(0)));
    };
    const auto load_ret = handle_load_reset_if_none<EventLabel::EventLabelKind::LockCasRead>(
        old_val_setter,
        thread_id,
        address,
        size,
        annot,
        EventDeps()
    );
    if (const auto* err = std::get_if<VerificationError>(&load_ret))
        return MutexLockResultExt::from_error(format_error(*err));

    const auto* ret_val = std::get_if<SVal>(&load_ret);
    if (!ret_val) {
        if (std::holds_alternative<Reset>(load_ret)) {
            // TODO TODO GENMC: what did I mean with this comment?
            // TODO GENMC: is_read_opt == Mutex is acquired
            // None	--> Someone else has lock, this thread will be rescheduled later
            // (currently block) 0	--> Got the lock 1 	--> Someone else has lock,
            // this thread will not be rescheduled later (block on Miri side)
            return MutexLockResultExt::ok(false);
        }
        ERROR("Unimplemented: mutex lock returned unexpected result.");
    }

    const bool is_lock_acquired = *ret_val == SVal(0);
    if (is_lock_acquired) {
        const auto store_ret = GenMCDriver::handleStore<EventLabel::EventLabelKind::LockCasWrite>(
            old_val_setter,
            inc_pos(thread_id),
            address,
            size,
            EventDeps()
        );

        if (const auto* err = std::get_if<VerificationError>(&store_ret))
            return MutexLockResultExt::from_error(format_error(*err));
        ERROR_ON(
            !std::holds_alternative<std::monostate>(store_ret),
            "Unsupported: mutex lock store returned unexpected result."
        );
    } else {
        ERROR_ON(*ret_val != SVal(1), "Mutex read value was neither 0 nor 1");
        GenMCDriver::handleAssume(inc_pos(thread_id), AssumeType::Spinloop);
    }

    return MutexLockResultExt::ok(is_lock_acquired);
}

auto MiriGenmcShim::handle_mutex_try_lock(ThreadId thread_id, uint64_t address, uint64_t size)
    -> MutexLockResult {
    const auto addr = SAddr(address);
    const auto aSize = ASize(size);

    auto& currPos = threads_action_[thread_id].event;
    // Mutex starts out unlocked, so we always say the previous value is "unlocked".
    auto old_val_setter = [this](SAddr addr) {
        this->handle_old_val(addr, GenmcScalarExt::from_sval(SVal(0)));
    };
    const auto load_ret = GenMCDriver::handleLoad<EventLabel::EventLabelKind::TrylockCasRead>(
        old_val_setter,
        ++currPos,
        addr,
        aSize
    );
    if (const auto* err = std::get_if<VerificationError>(&load_ret))
        return MutexLockResultExt::from_error(format_error(*err));
    const auto* ret_val = std::get_if<SVal>(&load_ret);
    if (nullptr == ret_val) {
        ERROR("Unimplemented: mutex trylock load returned unexpected result.");
    }

    const bool is_lock_acquired = *ret_val == SVal(0);
    if (!is_lock_acquired) {
        ERROR_ON(*ret_val != SVal(1), "Mutex read value was neither 0 nor 1");
        return MutexLockResultExt::ok(false); /* Lock already held. */
    }

    const auto store_ret = GenMCDriver::handleStore<EventLabel::EventLabelKind::TrylockCasWrite>(
        old_val_setter,
        ++currPos,
        addr,
        aSize
    );
    if (const auto* err = std::get_if<VerificationError>(&store_ret))
        return MutexLockResultExt::from_error(format_error(*err));
    if (!std::holds_alternative<std::monostate>(store_ret))
        ERROR("Unimplemented: mutex trylock store returned unexpected result.");
    // No error or unexpected result: lock is acquired.
    return MutexLockResultExt::ok(true);
}

auto MiriGenmcShim::handle_mutex_unlock(ThreadId thread_id, uint64_t address, uint64_t size)
    -> StoreResult {
    const auto pos = inc_pos(thread_id);
    const auto addr = SAddr(address);
    const auto aSize = ASize(size);

    const auto old_val_setter = [this](SAddr addr) {
        // TODO GENMC(HACK): is this the best way to do it?
        this->handle_old_val(addr, GenmcScalarExt::uninit());
    };
    const auto ret = GenMCDriver::handleStore<EventLabel::EventLabelKind::UnlockWrite>(
        old_val_setter,
        pos,
        MemOrdering::Release,
        addr,
        aSize,
        AType::Signed,
        SVal(0),
        EventDeps()
    );
    if (const auto* err = std::get_if<VerificationError>(&ret))
        return StoreResultExt::from_error(format_error(*err));
    if (!std::holds_alternative<std::monostate>(ret))
        ERROR("Unimplemented: mutex unlock store returned unexpected result.");

    // TODO GENMC: Mixed-accesses (`false` should be fine, since we never want to update Miri's
    // memory for mutexes anyway)
    // const bool is_coherence_order_maximal_write = false;
    const auto& g = getExec().getGraph();
    const bool is_coherence_order_maximal_write = g.co_max(addr)->getPos() == pos;
    return StoreResultExt::ok(is_coherence_order_maximal_write);
}
