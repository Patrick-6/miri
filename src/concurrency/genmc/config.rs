use super::GenmcParams;

/// Configuration for GenMC mode.
/// The `params` field is shared with the C++ side.
/// The remaining options are kept on the Rust side.
#[derive(Debug, Default, Clone)]
pub struct GenmcConfig {
    pub(super) params: GenmcParams,
    do_estimation: bool,
    // FIXME(GenMC): add remaining options.
}

impl GenmcConfig {
    /// Function for parsing command line options for GenMC mode.
    ///
    /// All GenMC arguments start with the string "-Zmiri-genmc".
    /// Passing any GenMC argument will enable GenMC mode.
    ///
    /// `trimmed_arg` should be the argument to be parsed, with the suffix "-Zmiri-genmc" removed.
    pub fn parse_arg(genmc_config: &mut Option<GenmcConfig>, trimmed_arg: &str) {
        // FIXME(genmc,macos): Add `target_os = "macos"` once `https://github.com/dtolnay/cxx/issues/1535` is fixed.
        if !cfg!(all(
            feature = "genmc",
            any(target_os = "linux", target_os = "macos"),
            target_pointer_width = "64",
            target_endian = "little"
        )) {
            unimplemented!(
                "GenMC mode is not supported on this platform, cannot handle argument: \"-Zmiri-genmc{trimmed_arg}\""
            );
        }
        if genmc_config.is_none() {
            *genmc_config = Some(Default::default());
        }
        if trimmed_arg.is_empty() {
            return; // this corresponds to "-Zmiri-genmc"
        }
        // FIXME(GenMC): implement remaining parameters.
        todo!();
    }
}
