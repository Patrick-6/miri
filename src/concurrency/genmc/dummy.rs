use rustc_abi::{Align, Size};
use rustc_const_eval::interpret::{AllocId, InterpCx, InterpResult};
use rustc_middle::mir;

use crate::{
    AtomicFenceOrd, AtomicReadOrd, AtomicRwOrd, AtomicWriteOrd, MemoryKind, MiriConfig,
    MiriMachine, OpTy, Scalar, ThreadId, ThreadManager, VisitProvenance, VisitWith,
};

#[derive(Debug)]
pub struct GenmcCtx {}

#[derive(Debug, Default, Clone)]
pub struct GenmcConfig {}

// TODO GENMC: add all exposed methods here too

impl GenmcCtx {
    pub fn new(_miri_config: &MiriConfig, _genmc_config: &GenmcConfig) -> Self {
        unreachable!()
    }
    
    pub fn print_estimation_result(&self) {
        unreachable!()
    }

    pub fn get_blocked_execution_count(&self) -> usize {
        unreachable!()
    }

    pub fn print_genmc_graph(&self) {
        unreachable!()
    }

    pub fn is_exploration_done(&self) -> bool {
        unreachable!()
    }

    /**** Memory access handling ****/

    pub(crate) fn handle_execution_start(&self) {
        unreachable!()
    }

    pub(crate) fn handle_execution_end<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
    ) -> Result<(), String> {
        unreachable!()
    }

    pub(super) fn set_ongoing_action_data_race_free(&self, _enable: bool) {
        unreachable!()
    }

    //* might fails if there's a race, load might also not read anything (returns None) */
    pub(crate) fn atomic_load<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _address: Size,
        _size: Size,
        _ordering: AtomicReadOrd,
        _old_val: Option<Scalar>,
    ) -> InterpResult<'tcx, Scalar> {
        unreachable!()
    }

    pub(crate) fn atomic_store<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _address: Size,
        _size: Size,
        _value: Scalar,
        _old_value: Option<Scalar>,
        _ordering: AtomicWriteOrd,
    ) -> InterpResult<'tcx, bool> {
        unreachable!()
    }

    pub(crate) fn atomic_fence<'tcx>(
        &self,
        _machine: &MiriMachine<'tcx>,
        _ordering: AtomicFenceOrd,
    ) -> InterpResult<'tcx, ()> {
        unreachable!()
    }

    pub(crate) fn atomic_rmw_op<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _address: Size,
        _size: Size,
        _ordering: AtomicRwOrd,
        (_rmw_op, _not): (mir::BinOp, bool),
        _rhs_scalar: Scalar,
        _old_value: Scalar,
    ) -> InterpResult<'tcx, (Scalar, Scalar)> {
        unreachable!()
    }

    pub(crate) fn atomic_min_max_op<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _address: Size,
        _size: Size,
        _ordering: AtomicRwOrd,
        _min: bool,
        _is_signed: bool,
        _rhs_scalar: Scalar,
        _old_value: Scalar,
    ) -> InterpResult<'tcx, (Scalar, Scalar)> {
        unreachable!()
    }

    pub(crate) fn atomic_exchange<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _address: Size,
        _size: Size,
        _rhs_scalar: Scalar,
        _ordering: AtomicRwOrd,
        _old_value: Scalar,
    ) -> InterpResult<'tcx, (Scalar, bool)> {
        unreachable!()
    }

    pub(crate) fn atomic_compare_exchange<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _address: Size,
        _size: Size,
        _expected_old_value: Scalar,
        _new_value: Scalar,
        _success: AtomicRwOrd,
        _fail: AtomicReadOrd,
        _can_fail_spuriously: bool,
        _old_value: Scalar,
    ) -> InterpResult<'tcx, (Scalar, bool)> {
        unreachable!()
    }

    pub(crate) fn memory_load<'tcx>(
        &self,
        _machine: &MiriMachine<'tcx>,
        _address: Size,
        _size: Size,
    ) -> InterpResult<'tcx, ()> {
        unreachable!()
    }

    pub(crate) fn memory_store<'tcx>(
        &self,
        _machine: &MiriMachine<'tcx>,
        _address: Size,
        _size: Size,
    ) -> InterpResult<'tcx, ()> {
        unreachable!()
    }

    /**** Memory (de)allocation ****/

    pub(crate) fn handle_alloc<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _alloc_id: AllocId,
        _size: Size,
        _alignment: Align,
        _memory_kind: MemoryKind,
    ) -> InterpResult<'tcx, u64> {
        unreachable!()
    }

    pub(crate) fn handle_dealloc<'tcx>(
        &self,
        _machine: &MiriMachine<'tcx>,
        _alloc_id: AllocId,
        _address: Size,
        _size: Size,
        _align: Align,
        _kind: MemoryKind,
    ) -> InterpResult<'tcx, ()> {
        unreachable!()
    }

    /**** Thread management ****/

    pub(crate) fn handle_thread_create<'tcx>(
        &self,
        _threads: &ThreadManager<'tcx>,
        _start_routine: crate::Pointer,
        _func_arg: &crate::ImmTy<'tcx>,
        _new_thread_id: ThreadId,
    ) -> InterpResult<'tcx, ()> {
        unreachable!()
    }

    pub(crate) fn handle_thread_join<'tcx>(
        &self,
        _active_thread_id: ThreadId,
        _child_thread_id: ThreadId,
    ) -> InterpResult<'tcx, ()> {
        unreachable!()
    }

    pub(crate) fn handle_thread_stack_empty<'tcx>(
        &self,
        _threads: &ThreadManager<'tcx>,
        _thread_id: ThreadId,
    ) {
        unreachable!()
    }

    pub(crate) fn handle_main_thread_stack_empty<'tcx>(&self, _threads: &ThreadManager<'tcx>) {
        unreachable!()
    }

    pub(crate) fn handle_thread_finish<'tcx>(&self, _threads: &ThreadManager<'tcx>) {
        unreachable!()
    }

    /**** Scheduling functionality ****/

    pub(crate) fn schedule_thread<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
    ) -> InterpResult<'tcx, ThreadId> {
        unreachable!()
    }
}

/// Other functionality not directly related to event handling
impl<'tcx> EvalContextExt<'tcx> for crate::MiriInterpCx<'tcx> {}
pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
    fn check_genmc_intercept_function(
        &mut self,
        _instance: rustc_middle::ty::Instance<'tcx>,
        _args: &[rustc_const_eval::interpret::FnArg<'tcx, crate::Provenance>],
        _dest: &crate::PlaceTy<'tcx>,
        _ret: Option<mir::BasicBlock>,
    ) -> InterpResult<'tcx, bool> {
        unreachable!()
    }

    /**** Blocking instructions ****/

    fn handle_genmc_verifier_assume(&mut self, _condition: &OpTy<'tcx>) -> InterpResult<'tcx> {
        unreachable!()
    }
}

impl VisitProvenance for GenmcCtx {
    fn visit_provenance(&self, _visit: &mut VisitWith<'_>) {
        unreachable!()
    }
}

impl GenmcConfig {
    pub fn parse_arg(_genmc_config: &mut Option<GenmcConfig>, trimmed_arg: &str) {
        unimplemented!(
            "GenMC feature im Miri is disabled, cannot handle argument: \"-Zmiri-genmc{trimmed_arg}\""
        );
    }

    pub fn print_exec_graphs(&self) -> bool {
        unreachable!()
    }

    pub fn do_estimation(&self) -> bool {
        unreachable!()
    }
}
