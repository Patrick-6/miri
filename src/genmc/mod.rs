use std::cell::{Cell, RefCell};

use rustc_abi::{Align, Size};
use rustc_const_eval::interpret::{InterpCx, InterpResult, interp_ok};

use self::helper::{
    NextInstrInfo, Threads, genmc_scalar_to_scalar, get_next_instr_info,
    rhs_scalar_to_genmc_scalar, scalar_to_genmc_scalar,
};
use self::thread_info_manager::{GenmcThreadId, GenmcThreadIdInner, ThreadInfoManager};
use crate::intrinsics::AtomicOp;
use crate::{
    AtomicFenceOrd, AtomicReadOrd, AtomicRwOrd, AtomicWriteOrd, MemoryKind, MiriConfig,
    MiriMachine, Scalar, TerminationInfo, ThreadId, ThreadManager, VisitProvenance, VisitWith,
};

mod config;

pub use self::config::GenmcConfig;
pub use self::ffi::GenmcParams;

// TODO: add fields
pub struct GenmcCtx {
    /// Some actions Miri does are allowed to cause data races.
    /// GenMC will not be informed about certain actions (e.g. non-atomic loads) when this flag is set
    allow_data_races: Cell<bool>,
}

impl GenmcCtx {
    /// Validate the selected configuration options and create a new `GenmcCtx` if successful
    ///
    /// Some combinations of options are (currently) not allowed:
    /// - Aliasing model checking is incompatible with GenMC mode
    ///   - The reason is that the required information is lost when pointers are send to GenMC and back
    /// - Data race checking and weak memory emulation must be turned off, since GenMC does this by itself
    /// - "Many seeds" mode in Miri is currently incompatible with GenMC mode
    pub fn try_new(_miri_config: &MiriConfig) -> Option<Self> {
        todo!()
    }

    pub fn get_stuck_execution_count(&self) -> usize {
        todo!()
    }

    pub fn print_genmc_graph(&self) {
        todo!()
    }

    pub fn is_exploration_done(&self) -> bool {
        todo!()
    }

    pub(crate) fn handle_execution_start(&self) {
        todo!()
    }

    pub(crate) fn handle_execution_end<'tcx>(
        &self,
        _thread_manager: &ThreadManager<'tcx>,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
    ) -> Result<(), String> {
        todo!()
    }

    /**** Memory access handling ****/

    /// If `true` is passed, allow for data races to happen without triggering an error, until this function is called again with argument `false`.
    /// This allows for racy non-atomic memory accesses to be ignored (GenMC is not informed about them at all).
    ///
    /// Certain operations are not permitted in GenMC mode with data races disabled and will cause a panic, e.g., atomic accesses or asking for scheduling decisions.
    ///
    /// # Panics
    /// This method will panic if data races are nested
    pub(crate) fn set_ongoing_action_data_race_free(&self, enable: bool) {
        let old = self.allow_data_races.replace(enable);
        assert_ne!(old, enable, "cannot nest allow_data_races");
    }

    pub(crate) fn atomic_load<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _address: Size,
        _size: Size,
        _ordering: AtomicReadOrd,
        _old_val: Option<Scalar>,
    ) -> InterpResult<'tcx, Scalar> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    pub(crate) fn atomic_store<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _address: Size,
        _size: Size,
        _value: Scalar,
        _ordering: AtomicWriteOrd,
    ) -> InterpResult<'tcx, ()> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    pub(crate) fn atomic_fence<'tcx>(
        &self,
        machine: &MiriMachine<'tcx>,
        ordering: AtomicFenceOrd,
    ) -> InterpResult<'tcx, ()> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    pub(crate) fn atomic_rmw_op<'tcx>(
        &self,
        ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        address: Size,
        size: Size,
        ordering: AtomicRwOrd,
        rmw_op: AtomicOp,
        rhs_scalar: Scalar,
        is_unsigned: bool,
    ) -> InterpResult<'tcx, (Scalar, bool)> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    pub(crate) fn atomic_exchange<'tcx>(
        &self,
        ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        address: Size,
        size: Size,
        rhs_scalar: Scalar,
        ordering: AtomicRwOrd,
    ) -> InterpResult<'tcx, (Scalar, bool)> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    pub(crate) fn atomic_compare_exchange<'tcx>(
        &self,
        ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        address: Size,
        size: Size,
        expected_old_value: Scalar,
        new_value: Scalar,
        success: AtomicRwOrd,
        fail: AtomicReadOrd,
        can_fail_spuriously: bool,
    ) -> InterpResult<'tcx, (Scalar, bool)> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    pub(crate) fn memory_load<'tcx>(
        &self,
        machine: &MiriMachine<'tcx>,
        address: Size,
        size: Size,
    ) -> InterpResult<'tcx, Scalar> {
        todo!()
    }

    pub(crate) fn memory_store<'tcx>(
        &self,
        machine: &MiriMachine<'tcx>,
        address: Size,
        size: Size,
    ) -> InterpResult<'tcx, ()> {
        todo!()
    }

    /**** Memory (de)allocation ****/

    pub(crate) fn handle_alloc<'tcx>(
        &self,
        machine: &MiriMachine<'tcx>,
        size: Size,
        alignment: Align,
        memory_kind: MemoryKind,
    ) -> InterpResult<'tcx, u64> {
        todo!()
    }

    pub(crate) fn handle_dealloc<'tcx>(
        &self,
        machine: &MiriMachine<'tcx>,
        address: Size,
        size: Size,
        align: Align,
        kind: MemoryKind,
    ) -> InterpResult<'tcx, ()> {
        todo!()
    }

    /**** Thread management ****/

    pub(crate) fn handle_thread_create<'tcx>(
        &self,
        threads: &ThreadManager<'tcx>,
        new_thread_id: ThreadId,
    ) -> InterpResult<'tcx, ()> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    pub(crate) fn handle_thread_join(
        &self,
        active_thread_id: ThreadId,
        child_thread_id: ThreadId,
    ) -> Result<(), ()> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    pub(crate) fn handle_thread_stack_empty(&self, _thread_id: ThreadId) {
        todo!()
    }

    pub(crate) fn handle_thread_finish<'tcx>(
        &self,
        threads: &ThreadManager<'tcx>,
    ) -> InterpResult<'tcx, ()> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    /**** Scheduling functionality ****/

    pub(crate) fn should_preempt(&self) -> bool {
        true // TODO GENMC
    }

    pub(crate) fn schedule_thread<'tcx>(
        &self,
        thread_manager: &ThreadManager<'tcx>,
        ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
    ) -> InterpResult<'tcx, ThreadId> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    /**** Blocking instructions ****/

    pub(crate) fn handle_verifier_assume<'tcx>(
        &self,
        machine: &MiriMachine<'tcx>,
        condition: bool,
    ) -> InterpResult<'tcx, ()> {
        if condition { interp_ok(()) } else { self.handle_user_block(machine) }
    }
}

impl VisitProvenance for GenmcCtx {
    fn visit_provenance(&self, _visit: &mut VisitWith<'_>) {
        // We don't have any tags.
    }
}

impl GenmcCtx {
    fn handle_user_block<'tcx>(&self, machine: &MiriMachine<'tcx>) -> InterpResult<'tcx, ()> {
        todo!()
    }
}
