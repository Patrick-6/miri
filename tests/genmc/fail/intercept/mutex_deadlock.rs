//@compile-flags: -Zmiri-genmc -Zmiri-disable-stacked-borrows
//@error-in-other-file: unsupported operation

// TODO

#![no_main]
#![feature(abort_unwind)]

#[path = "../../../utils/genmc.rs"]
mod genmc;

use std::sync::Mutex;

use crate::genmc::*;

static X: Mutex<u64> = Mutex::new(0);
static Y: Mutex<u64> = Mutex::new(0);

#[unsafe(no_mangle)]
fn miri_start(_argc: isize, _argv: *const *const u8) -> isize {
    unsafe {
        let t0 = spawn_pthread_closure(|| {
            let mut x = X.lock().unwrap();
            let mut y = Y.lock().unwrap();
            *x += 1;
            *y += 1;
        });
        let t1 = spawn_pthread_closure(|| {
            let mut y = Y.lock().unwrap();
            let mut x = X.lock().unwrap();
            *x += 1;
            *y += 1;
        });
        join_pthreads([t0, t1]);
        0
    }
}
