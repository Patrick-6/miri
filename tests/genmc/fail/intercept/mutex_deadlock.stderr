Running GenMC Verification...
error: unsupported operation: Could not lock Mutex, which may indicate a deadlock. (GenMC mode does not support deadlock detection yet).
  --> RUSTLIB/std/src/sync/poison/mutex.rs:LL:CC
   |
LL |             self.inner.lock();
   |                             ^ unsupported operation occurred here
   |
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that Miri does not support
   = note: BACKTRACE on thread `unnamed-ID`:
   = note: inside `std::sync::Mutex::<u64>::lock` at RUSTLIB/std/src/sync/poison/mutex.rs:LL:CC
note: inside closure
  --> tests/genmc/fail/intercept/mutex_deadlock.rs:LL:CC
   |
LL |             let mut x = X.lock().unwrap();
   |                         ^^^^^^^^
   = note: inside `<std::boxed::Box<{closure@tests/genmc/fail/intercept/mutex_deadlock.rs:LL:CC}> as std::ops::FnOnce<()>>::call_once` at RUSTLIB/alloc/src/boxed.rs:LL:CC
note: inside `genmc::spawn_pthread_closure::thread_func::<{closure@tests/genmc/fail/intercept/mutex_deadlock.rs:LL:CC}>`
  --> tests/genmc/fail/intercept/../../../utils/genmc.rs:LL:CC
   |
LL |         f();
   |         ^^^

note: add `-Zmiri-genmc-print-genmc-output` to MIRIFLAGS to see the detailed GenMC error report

error: aborting due to 1 previous error

